tests/test_moisture_sensor.py

```diff
@@ -7,21 +7,38 @@
 
 class MoistureSensorTest(unittest.TestCase):
 
-    def test_moisture_50(self):
+    def test_moisture_read_succeeds_when_adc_succeeds(self):
         """Moisture level read should be the average of the two readings."""
         mock_adc = mock.Mock()
         mock_adc.read_adc.side_effect = [300, (1023 - 400)]
         mock_pi_io = mock.Mock()
         mock_clock = mock.Mock()
         channel = 1
-        gpio_pins = (12, 16)
-        sensor = moisture_sensor.MoistureSensor(mock_adc, mock_pi_io, channel,
-                                                gpio_pins, mock_clock)
-        moisture_level = sensor.moisture()
-
-        self.assertIn(mock.call(12), mock_pi_io.turn_pin_on.call_args_list)
-        self.assertIn(mock.call(12), mock_pi_io.turn_pin_off.call_args_list)
-        self.assertIn(mock.call(16), mock_pi_io.turn_pin_on.call_args_list)
-        self.assertIn(mock.call(16), mock_pi_io.turn_pin_off.call_args_list)
-        mock_clock.wait.assert_called_once()
-        self.assertEqual(moisture_level, 350)
+        gpio_pin_1 = 12
+        gpio_pin_2 = 16
+        sensor = moisture_sensor.MoistureSensor(
+            mock_adc, mock_pi_io, channel, gpio_pin_1, gpio_pin_2, mock_clock)
+
+        self.assertEqual(350, sensor.moisture())
+        self.assertEqual([mock.call(12), mock.call(16)],
+                         mock_pi_io.turn_pin_on.call_args_list)
+        self.assertEqual(
+            [mock.call(16), mock.call(12), mock.call(12), mock.call(16)],
+            mock_pi_io.turn_pin_off.call_args_list)
+        mock_clock.wait.assert_called_once_with(0.1)
+
+    def test_moisture_sensor_pins_turned_off_if_adc_raises_exception(self):
+        mock_adc = mock.Mock()
+        mock_adc.read_adc.side_effect = ValueError()
+        mock_pi_io = mock.Mock()
+        mock_clock = mock.Mock()
+        channel = 1
+        gpio_pin_1 = 12
+        gpio_pin_2 = 16
+        sensor = moisture_sensor.MoistureSensor(
+            mock_adc, mock_pi_io, channel, gpio_pin_1, gpio_pin_2, mock_clock)
+
+        with self.assertRaises(ValueError):
+            sensor.moisture()
+            self.assertEqual([mock.call(16), mock.call(12), mock.call(16)],
+                             mock_pi_io.turn_pin_off.call_args_list)
```

Comment:
Oh yeah! We can do it in `setUp`. I was thinking about trying to extract just the IO stuff to its own separate helper class and didn't think about `setUp` but you're right. This should work:

```python
    def setUp(self):
        self.mock_adc = mock.Mock()
        self.mock_adc.read_adc.side_effect = ValueError()
        self.pin_state = collections.defaultdict(lambda: False)

        def mock_turn_pin_on(pin):
            self.pin_state[pin] = True

        def mock_turn_pin_off(pin):
            self.pin_state[pin] = False

        self.mock_pi_io = mock.Mock()
        self.mock_pi_io.turn_pin_on.side_effect = mock_turn_pin_on
        self.mock_pi_io.turn_pin_off.side_effect = mock_turn_pin_off
        self.mock_clock = mock.Mock()
```

>Can you reiterate for me why you don't like using setUp? I think the idea is that someone might only be interested in reading the code for some tests, and not others, and it makes it more difficult to have to scroll to setUp to figure out what's going on? 

Oh, I don't dislike `setUp` in general. I just think we need to be deliberate about what we put in it. Like [`test_poller.py`](https://github.com/JeetShetty/GreenPiThumb/blob/master/tests/test_poller.py) I think uses `setUp` well. But if we changed it to be like:

```python
    def setUp(self):
        self.clock_wait_event = threading.Event()
        self.mock_local_clock = mock.Mock()
        self.mock_sensor = mock.Mock()
        self.mock_store = mock.Mock()
        self.mock_sensor.temperature.return_value = 21.0 # <<< Bad

    def test_temperature_poller(self):
        temperature_poller = poller.TemperaturePoller(
            self.mock_local_clock, POLL_INTERVAL, self.mock_sensor,
            self.mock_store)
        self.mock_local_clock.now.return_value = TIMESTAMP_A
        self.mock_local_clock.wait.side_effect = (
            lambda _: self.clock_wait_event.set())

        temperature_poller.start_polling_async()
        self.clock_wait_event.wait(TEST_TIMEOUT_SECONDS)
        # BAD: Where is 21.0 coming from?
        self.mock_store.store_temperature.assert_called_with(TIMESTAMP_A, 21.0)
        self.mock_local_clock.wait.assert_called_with(POLL_INTERVAL)
```

Then it's confusing because it buries where the `21.0` is coming from so the reader has to search around.

`setUp` is good for building up scaffolding for the tests where the particulars don't matter much. It's bad for setting particular values that matter in the test.